ract/cubes/_gpio_impl.py

```diff
@@ -13,38 +13,66 @@
 
 
 class GpioCube(object):
-    """ Control the tesseract by directly toggling pins using GPIO. """
+    """ Raspberry Pi 1 hardware abstraction layer (HAL) for controlling the tesseract by directly toggling pins using
+    GPIO.
+    """
 
     def __init__(self):
-        pass
+        """ Initialization
+        """
+        self._fps = None
+        self._flattened_frame_layers = None
+        self._tesseract = None
 
     def setup(self, fps):
+        """ Set up the cube hardware abstraction layer, which means:
+         - instantiate a tesseract interface
+         - set up the GPIO interface
+         - clock in the LED dot-correction data
+        :param fps:
+        :return:
+        """
         faulthandler.enable()
         self._tesseract = GpioTesseract()
         self._tesseract.setup()
+        self._tesseract.clock_in_dot_correction(self._tesseract.default_dot_correction())
         self._fps = fps
+        self._flattened_frame_layers = None
 
     def teardown(self):
-        self._tesseract.teardown()
+        """ Tear down the GPIO interface if it has been setup
+        :return:
+        """
+        if self._tesseract:
+            self._tesseract.teardown()
 
     def render(self, frame):
-        pass
+        """ Render data into a suitable format for the GPIO input
+        :param frame:
+        :return:
+        """
+        self._flattened_frame_layers = []
+        for layer in frame:
+            self._flattened_frame_layers.append(layer.reshape(64))
```

Comment:
Yeah, was thinking that. Wasn't sure what it would do to the data structure, but will test it.